src/lib/API/Repository/Value/Filter/Filter.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Platform\API\Repository\Value\Filter;
+
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+use eZ\Publish\Core\Base\Exceptions\BadStateException;
+
+/**
+ * Content & Location filtering input Filter.
+ */
+final class Filter
+{
+    /**
+     * @internal use LocationService::createFindFilter() or ContentService::createFindFilter() instead.
+     *
+     * @see \eZ\Publish\API\Repository\LocationService::createFindFilter()
+     * @see \eZ\Publish\API\Repository\ContentService::createFindFilter()
+     */
+    public function __construct()
+    {
+    }
+
+    /**
+     * Reset Filter so it can be built from scratch.
+     */
+    public function reset(): void
+    {
+        $this->criterion = $this->sortClauses = null;
+        $this->languageCodes = [];
+    }
+
+    /**
+     * Set filtering Criterion.
+     *
+     * If multiple Criteria are required, either use `andWithCriterion`/`orWithCriterion` or wrap
+     * them with Logical operator Criterion.
+     *
+     * To re-build Criterion from scratch `reset` it first.
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException if Criterion is already set
+     *
+     * @see \Ibexa\Platform\API\Repository\Value\Filter\Filter::reset
+     * @see \Ibexa\Platform\API\Repository\Value\Filter\Filter::andWithCriterion
+     * @see \Ibexa\Platform\API\Repository\Value\Filter\Filter::orWithCriterion
+     * @see \eZ\Publish\API\Repository\Values\URL\Query\Criterion\LogicalOr
+     * @see \eZ\Publish\API\Repository\Values\URL\Query\Criterion\LogicalAnd
+     */
+    public function withCriterion(Criterion $criterion): Filter
+    {
+        if (null !== $this->criterion) {
+            throw new BadStateException('$criterion', 'Criterion is already set');
+        }
+
+        $this->criterion = $criterion;
+
+        return $this;
+    }
+
+    /**
+     * @see \Ibexa\Platform\API\Repository\Value\Filter\Filter::withCriterion
+     */
+    public function andWithCriterion(Criterion $criterion): Filter
+    {
+        $this->criterion = null !== $this->criterion
+            ? new Criterion\LogicalAnd([$this->criterion, $criterion])
```

Comment:
If `$this->criterion` is already an instance of `Criterion\LogicalAnd`, it should add the given criterion to it.

Reply:
> If `$this->criterion` is already an instance of `Criterion\LogicalAnd`, it should add the given criterion to it.

yeah, I felt there will be an issue with this simple expression. Actually working on unit tests now to visualize better the way complex Filter is built.